server/matchmaker/algorithm/team_matchmaker.py

```diff
@@ -0,0 +1,314 @@
+        self._logger.debug("Running Karmarkar-Karp to partition the teams")
+        # Further reading: https://en.wikipedia.org/wiki/Largest_differencing_method
+        # Karmarkar-Karp works only for positive integers. By adding 5000 to the rating of each player
+        # we also strongly incentivise the algorithm to give both teams the same number of players
+        containers = SortedList([Container(5000 * len(s.players) + s.cumulated_rating, [s]) for s in searches],
+                                key=lambda c: c.rating)
+
+        elem1 = containers.pop()
+        elem2 = containers.pop()
+        while True:
+            #  elem1 is always bigger than elem2
+            container = Container(elem1.rating - elem2.rating, [elem1, elem2])
+            containers.add(container)
+            elem1 = containers.pop()
+            try:
+                elem2 = containers.pop()
+            except IndexError:
+                break
+        self._logger.debug("Rating disparity: %s", elem1.rating)
+
+        # We now need to open all containers again to get to the searches. A container can hold a single
+        # search or two other containers, so we differentiate the two cases by the length of the content array.
+        # Because each container represent the difference of the two containing elements they have to go in
+        # different teams (The higher one into the team the container is in).
+        team_a = []
+        team_b = []
+        containers_a = []
+        containers_b = []
+        containers_a.append(elem1.content[0])
+        containers_b.append(elem1.content[1])
+        while containers_a or containers_b:
+            for e in containers_a:
+                if len(e.content) == 2:
+                    containers_a.append(e.content[0])
+                    containers_b.append(e.content[1])
+                else:
+                    team_a.append(e.content[0])
+                containers_a.remove(e)
+            for e in containers_b:
+                if len(e.content) == 2:
+                    containers_b.append(e.content[0])
+                    containers_a.append(e.content[1])
+                else:
+                    team_b.append(e.content[0])
+                containers_b.remove(e)
+        return team_a, team_b
+
+    def _searches_by_size(self, searches: List[Search]) -> Dict[int, List[Search]]:
+        searches_by_size: Dict[int, List[Search]] = defaultdict(list)
+
+        for search in searches:
+            searches_by_size[len(search.players)].append(search)
+
+        if self._logger.isEnabledFor(logging.DEBUG):
+            max_size = max(searches_by_size.keys())
+            self._logger.debug("participating searches by player size:")
+            for i in range(1, max_size + 1):
+                self._logger.debug("%i players: %s" % (i, searches_by_size[i]))
+        return searches_by_size
+
+    def _find_most_balanced_filler(self, avg, search, searches_dict):
```

Comment:
I don't think you need to pass the whole `search` or the whole `searches_dict`, isn't this simply finding the player in `searches_dict[1]` whose rating is closest to the target of `team_size * avg - get_cumulative_rating(search)`? Then I would pass that number and `searches_dict[1]` instead.